pkg/shared/clients/redis/redis_reader.go

```diff
@@ -0,0 +1,150 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package redis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/numaproj/numaflow/pkg/isb"
+	"go.uber.org/zap"
+)
+
+// RedisStreamsReader is the read queue implementation powered by RedisClient.
+type RedisStreamsReader struct {
+	Name     string
+	Stream   string
+	Group    string
+	Consumer string
+
+	*RedisClient
+	Options
+	Log     *zap.SugaredLogger
+	Metrics Metrics
+
+	XStreamToMessages func(xstreams []redis.XStream, messages []*isb.ReadMessage, labels map[string]string) ([]*isb.ReadMessage, error)
+}
+
+type Metrics struct {
+	ReadErrorsInc MetricsIncrementFunc
+	ReadsAdd      MetricsAddFunc
+	AcksAdd       MetricsAddFunc
+}
+
+// need a function type which increments a particular counter
+type MetricsIncrementFunc func()
+type MetricsAddFunc func(int)
+
+func (br *RedisStreamsReader) GetName() string {
+	return br.Name
+}
+
+// GetStreamName returns the stream name.
+func (br *RedisStreamsReader) GetStreamName() string {
+	return br.Stream
+}
+
+// GetGroupName gets the name of the consumer group.
+func (br *RedisStreamsReader) GetGroupName() string {
+	return br.Group
+}
+
+// Read reads the messages from the stream.
+// During a restart, we need to make sure all the un-acknowledged messages are reprocessed.
+// we need to replace `>` with `0-0` during restarts. We might run into data loss otherwise.
+func (br *RedisStreamsReader) Read(_ context.Context, count int64) ([]*isb.ReadMessage, error) {
+	var messages = make([]*isb.ReadMessage, 0, count)
+	var xstreams []redis.XStream
+	var err error
+	// start with 0-0 if CheckBackLog is true
+	labels := map[string]string{"buffer": br.GetName()}
+	if br.Options.CheckBackLog {
+		xstreams, err = br.processXReadResult("0-0", count)
+		if err != nil {
+			return br.processReadError(xstreams, messages, err)
+		}
+
+		// NOTE: If all messages have been delivered and acknowledged, the XREADGROUP 0-0 call returns an empty
+		// list of messages in the stream. At this point we want to read everything from last delivered which would be >
+		if len(xstreams) == 1 && len(xstreams[0].Messages) == 0 {
+			br.Log.Infow("We have delivered and acknowledged all PENDING msgs, setting checkBacklog to false")
+			br.CheckBackLog = false
+		}
+	}
+	if !br.Options.CheckBackLog {
+		xstreams, err = br.processXReadResult(">", count)
+		if err != nil {
+			return br.processReadError(xstreams, messages, err)
+		}
+	}
+
+	// for each XMessage in []XStream
+	msgs, err := br.XStreamToMessages(xstreams, messages, labels)
+	if br.Metrics.ReadsAdd != nil {
+		br.Metrics.ReadsAdd(len(messages))
+	}
+	br.Log.Debugf("received %d messages over Redis Streams Source, err=%v", len(msgs), err)
```

Comment:
Capital first letter for logging. `Received ...`

Reply:
I see. Is it capital letter for standard logging and lower case for errors? (I noticed lint had an issue when I capitalized an error)